lib/evaluate/check-expression.cc

```diff
@@ -0,0 +1,240 @@
+  }
+  Result operator()(const CoarrayRef &) const { return "coindexed reference"; }
+
+  Result operator()(const semantics::Symbol &symbol) const {
+    if (semantics::IsNamedConstant(symbol)) {
+      return std::nullopt;
+    } else if (symbol.IsDummy()) {
+      if (symbol.attrs().test(semantics::Attr::OPTIONAL)) {
+        return "reference to OPTIONAL dummy argument '"s +
+            symbol.name().ToString() + "'";
+      } else if (symbol.attrs().test(semantics::Attr::INTENT_OUT)) {
+        return "reference to INTENT(OUT) dummy argument '"s +
+            symbol.name().ToString() + "'";
+      } else if (symbol.has<semantics::ObjectEntityDetails>()) {
+        return std::nullopt;
+      } else {
+        return "dummy procedure argument";
+      }
+    } else if (symbol.has<semantics::UseDetails>() ||
+        symbol.has<semantics::HostAssocDetails>() ||
+        symbol.owner().kind() == semantics::Scope::Kind::Module) {
+      return std::nullopt;
+    } else if (const auto *object{
+                   symbol.detailsIf<semantics::ObjectEntityDetails>()}) {
+      // TODO: what about EQUIVALENCE with data in COMMON?
+      // TODO: does this work for blank COMMON?
+      if (object->commonBlock() != nullptr) {
+        return std::nullopt;
+      }
+    }
+    for (const semantics::Scope *s{&scope_}; !s->IsGlobal();) {
+      s = &s->parent();
+      if (s == &symbol.owner()) {
+        return std::nullopt;
+      }
+    }
+    return "reference to local entity '"s + symbol.name().ToString() + "'";
+  }
+
+  Result operator()(const Component &x) const {
+    // Don't look at the component symbol.
+    return (*this)(x.base());
+  }
+  Result operator()(const DescriptorInquiry &) const {
+    // Subtle: Uses of SIZE(), LBOUND(), &c. that are valid in specification
+    // expressions will have been converted to expressions over descriptor
+    // inquiries by Fold().
+    return std::nullopt;
+  }
+
+  template<typename T> Result operator()(const FunctionRef<T> &x) const {
+    if (const auto *symbol{x.proc().GetSymbol()}) {
+      if (!symbol->attrs().test(semantics::Attr::PURE)) {
+        return "reference to impure function '"s + symbol->name().ToString() +
+            "'";
+      }
+      // TODO: other checks for standard module procedures
+    } else {
+      const SpecificIntrinsic &intrin{DEREF(x.proc().GetSpecificIntrinsic())};
+      if (intrin.name == "present") {
+        return std::nullopt;  // no need to check argument(s)
+      }
+      if (IsConstantExpr(x)) {
+        // inquiry functions may not need to check argument(s)
+        return std::nullopt;
+      }
+    }
+    return (*this)(x.arguments());
+  }
+
+private:
+  const semantics::Scope &scope_;
+};
+
+template<typename A>
+void CheckSpecificationExpr(const A &x, parser::ContextualMessages &messages,
+    const semantics::Scope &scope) {
+  if (auto why{CheckSpecificationExprHelper{scope}(x)}) {
+    std::stringstream ss;
+    ss << x;
```

Comment:
Is it possible to use `AsFortran()` here? That's more specific about what kind of string representation will come out.

Reply:
I could use `x.AsFortran(ss)` but I don't think that the output would be different.

Reply:
Also, `operator<<` copes with `std::optional<>` better.

Reply:
`operator<<` can handle `std::optional<>`.